cmd/server/main.go

```diff
@@ -77,3 +66,60 @@ func runServer(ctx context.Context, addr string, router *chi.Mux) {
 	<-shutdownCtx.Done()
 	log.Println("server stopped")
 }
+
+func setupRouter(storage storage.Storage, logger *server.Logger, db *sql.DB) *chi.Mux {
+	mvLogger := middleware.NewHTTPLogger(logger)
+	updaterHandler := update.New(storage, logger)
+	updaterJSONHandler := update.NewJSONHandler(storage, logger)
+	updaterBatchHandler := updates.New(storage, logger)
+	retrieverHandler := retrieve.New(storage, logger)
+	retrieverJSONHandler := retrieve.NewJSONHandler(storage, logger)
+	pingHandler := ping.New(db)
+
+	router := chi.NewRouter()
+	router.Use(mvLogger.Log)
+	router.Use(middleware.GzipWrapper)
+
+	router.Route("/update", func(r chi.Router) {
+		r.Post("/{type}/{name}/{value}", updaterHandler.Update)
+		r.Post("/", updaterJSONHandler.Update)
+	})
+	router.Route("/updates", func(r chi.Router) {
+		r.Post("/", updaterBatchHandler.Update)
+	})
+	router.Route("/value", func(r chi.Router) {
+		r.Get("/{type}/{name}", retrieverHandler.Retrieve)
+		r.Post("/", retrieverJSONHandler.Retrieve)
+	})
+	router.Get("/ping", pingHandler.Ping)
+	router.Get("/", info.InfoPage)
+
+	return router
+}
+
+func mustSetupStorage(config *server.Config, logger *server.Logger) (storage.Storage, *sql.DB) {
+	db, err := sql.Open("pgx", config.DatabaseDSN)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if config.DatabaseDSN != "" {
+		if err := db.Ping(); err != nil {
+			log.Fatal(err)
+		}
+
+		storage, err := pg.New(db)
+		if err != nil {
+			log.Fatal(err)
+		}
+
+		return storage, db
+	}
+
+	storage, err := file.New(config.FileCfg, logger)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	return storage, db
+}
```

Comment:
Я не рекомендую вызывать `log.Fatal(...)` из функций или компонентов кроме метода `main(...)`. Даже если кажется, что это приемлемо в случае вызова в том же файле - на мой взгляд с точки зрения читаемости стоит обработать ошибку, добавить к ней дополнительный контекст, если требуется и вернуть в main для обработки.
```suggestion
func mustSetupStorage(config *server.Config, logger *server.Logger) (storage.Storage, *sql.DB, error) {
	db, err := sql.Open("pgx", config.DatabaseDSN)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to open database: %w", err)
	}

	if config.DatabaseDSN != "" {
		if err := db.Ping(); err != nil {
			return nil, nil, fmt.Errorf("database is not reachable: %w", err)
		}

		storage, err := pg.New(db)
		if err != nil {
			return nil, nil, fmt.Errorf("failed to create storage: %w", err)
		}

		return storage, db, nil
	}

	storage, err := file.New(config.FileCfg, logger)
	if err != nil {
		return nil, nil, fmt.Errorf("failed to create storage: %w", err)
	}

	return storage, db, nil
}
```

Reply:
Выглядит логично, но читабельности коду это точно не добавляет.
Но я на одном посвященном Go ютуб канале, что есть такое соглашение: если функция инициализации имеет приставку must (MustLoad, mustSetup), то предполагается, что эта функция может завершить приложение. Кроме того на других  похожих каналах, я видел завершение приложения из пакета конфига (log.Fatal(err)) в случае неудачно реализации.
Что мне делать с функцией runServer? Там не получиться вернуть ошибку, так как она возникает в горутине.